File: OLD/cpp/src/bubble-sort.cpp
#include "../headers/Arrays.h"
#include "../headers/Time.h"

#include <cmath>

using std::cout;

std::shared_ptr<int[]> _bubble_sort(std::shared_ptr<int[]> arr, const int size)
{
    for (int i = 0; i < size; i++) {
        for (int j = size - 1; j > i; j--) {
            if (arr[j - 1] > arr[j]) {
                std::swap(arr[j - 1], arr[j]);
            }
        }
    }
    return arr;
}

Result bubble_sort(const std::shared_ptr<int[]> arr, const int size)
{
    auto arr_copy = copy_array(arr, size);
    auto start = clock();
    auto res = _bubble_sort(arr_copy, size);
    auto end = clock();
    auto sortTime = get_time_ms(start, end);
    return (Result) {res, sortTime};
}

int main()
{
    const int size = 40000;

    // Generate shuffled array
    auto genResult = generate_shuffled_array(size);

    // Sort shuffled array
    auto sortResult = bubble_sort(genResult.arr, size);

    // Eval sorted array with an expected result
    auto expected = generate_expected_array(size);

    // Output the results
    cout << "1. Time to generate array of size " << size << ": "
         << floor(genResult.time) << " ms\n";
    if (! is_sorted_as_expected(sortResult.arr, expected, size)) {
        cout << "2. The result array is NOT sorted as expected\n";
    } else {
        cout << "2. The result array is sorted as expected\n";
        cout << "3. Time to bubble sort the array: "
             << floor(sortResult.time) << " ms\n";
    }

    return 0;
}

File: OLD/javascript/bubble-sort.js
const { generateShuffledArray, generateExpectedArray } =  require("./arrays")

const _bubbleSort = (arr) =>
{
    const res = arr.slice(0)
    for (let i = 0; i < res.length; i++) {
        let swapCount = 0
        for (let j = res.length - 1; j > i; j--) {
            if (res[j] < res[j - 1]) {
                // Swap
                const tmp = res[j]
                res[j] = res[j - 1]
                res[j - 1] = tmp
                swapCount++
            }
        }
        // Optimization: Stop looping after array already sorted
        // *Only logical: actually keeping the extra variable makes it a little slower
        if (swapCount == 0) break
    }
    return res
}

const bubbleSort = (arr) =>
{
    const start = new Date().getTime()
    const resultArray = _bubbleSort(arr)
    const end = new Date().getTime()
    return {
        res: resultArray,
        sortTime: end - start
    }
}

const main = () =>
{
    const size = 40000

    // Generate array
    const { arr, genTime } = generateShuffledArray(size)

    // Sort shuffled array
    const { res, sortTime } = bubbleSort(arr)

    // Eval result array with an expected result
    const expected = generateExpectedArray(size)
    const isSortedAsExpected = res.toString() == expected.toString()

    // Output the results
    console.log(`Time to generate shuffled array of size ${size}: ${genTime} ms`)
    if (! isSortedAsExpected) {
        console.log("2. The result arrays is NOT sorted as expected")
    } else {
        console.log("2. The result arrays is sorted as expected")
        console.log(`3. Time to bubble sort the shuffled array: ${sortTime} ms`)
    }
}

main()
